### ISSUE TITLE
Synology DSM NAS nextdns wont start after install

### ISSUE THREAD

**Issue by jbehrmusic** [nextdns/nextdns#420]
The install seemed to be fine. But when I try to run **nextdns status**, I get this.

~# nextdns status
Error: /usr/local/etc/rc.d/nextdns.sh status: exit status 1:

when I try to run **nextdns run**, I get this.

~# nextdns version
nextdns version 1.10.1
~# nextdns run
INFO: 05:07:51 Starting NextDNS 1.10.1/linux on localhost:53
INFO: 05:07:51 Listening on TCP/[::1]:53
INFO: 05:07:51 Listening on TCP/127.0.0.1:53
INFO: 05:07:51 Listening on UDP/[::1]:53
INFO: 05:07:51 Listening on UDP/127.0.0.1:53
ERROR: 05:07:51 Startup failed: proxy: tcp: listen tcp [::1]:53: bind: address already in use
Error: proxy: tcp: listen tcp [::1]:53: bind: address already in use

**Lastly, here's the log:**

~# nextdns log
2021-02-12T04:09:10-08:00 nextdns[4288]: Cannot start control server: listen unix /var/run/nextdns.sock: bind: address already in use
2021-02-12T04:09:11-08:00 nextdns[4288]: Startup failed: proxy: tcp: listen tcp :53: bind: permission denied
2021-02-12T04:30:59-08:00 nextdns[9170]: Startup failed: proxy: tcp: listen tcp :53: bind: address already in use
2021-02-12T04:31:20-08:00 nextdns[9317]: Startup failed: proxy: tcp: listen tcp :53: bind: address already in use
2021-02-12T05:05:14-08:00 nextdns[27810]: Startup failed: proxy: tcp: listen tcp :53: bind: address already in use
2021-02-12T05:07:26-08:00 nextdns[27936]: Startup failed: proxy: tcp: listen tcp :53: bind: address already in use
2021-02-12T05:09:47-08:00 nextdns[28193]: Startup failed: proxy: tcp: listen tcp :53: bind: address already in use

**Comment by rs** [comment]
Don’t use « run » but « start » or « restart ».

**Comment by dgw** [comment]
Should probably also see what's already on port 53: `lsof -i:53`

**Comment by rs** [comment]
> Should probably also see what's already on port 53: `lsof -i:53`

It's probably `nextdns` as `run` is what `start` use from the init script. If you use `run` while the service is already installed, you just start a new instance.

**Comment by dgw** [comment]
If it's nextdns, it confirms the theory. If the system already has another resolver running, it will reveal that too. Win-win!

**Comment by jbehrmusic** [comment]
Ok. Using the restart command, then the start command seemed to allow the service to run. When I type "status", it says "running". So that seems to be good. But when I run the log again, this is what I get.

2021-02-12T13:44:45-08:00 nextdns[13413]: Cannot start control server: listen unix /var/run/nextdns.sock: bind: permission denied
2021-02-12T13:44:45-08:00 nextdns[13413]: Startup failed: proxy: tcp: listen tcp :53: bind: permission denied
2021-02-12T13:45:11-08:00 nextdns[13472]: Cannot start control server: listen unix /var/run/nextdns.sock: bind: permission denied
2021-02-12T13:45:11-08:00 nextdns[13472]: Startup failed: proxy: tcp: listen tcp :53: bind: permission denied

So it seems like a permissions issue? I also tried running the lsof command, but I got this.

~# lsof -i:53
-ash: lsof: command not found

**Comment by dgw** [comment]
If the error's changed from "address in use" to "permission denied", then you're probably fine as far as conflicting services go. (Busybox should have `lsof` if you install it, or try using `netstat -lp`.)

The first 1024 ports on most \*nix systems are restricted to use by services running with root permissions only. Synology's system might have some other way for processes to gain access to those lower ports without running as root, but I'm not familiar with their environment.

**Comment by jbehrmusic** [comment]
not sure how to get busybox installed on my nas. and netstat -lp doesn't show anything regarding port 53.

**Comment by jbehrmusic** [comment]
Update: For some reason, everything seems to be working now. I guess after I restarted my NAS and ran the install again (uninstalled/then re-installed), it looks like it is working properly.

**Comment by netfxtech** [comment]
Jbehr- It’s not you. There’s something wrong with NextDNS on Synology. I manager 17 Synology’s and 7 run nextdns. All of them are having this same issue. It just started yesterday or the day before. I’ll start troubleshooting this weekend.

**Comment by netfxtech** [comment]
If it helps in the meantime, all of them show nextdns is running when I

netstat-tulpn | grep 53

I’ve tried killing the service, restart service, reboot device, uninstall/reinstall. None of them are resolving DNS.

**Comment by netfxtech** [comment]
> Update: For some reason, everything seems to be working now. I guess after I restarted my NAS and ran the install again (uninstalled/then re-installed), it looks like it is working properly.
> 
> The question is. Should I still be scared of that log line, which says the **bind address is already in use**?
> 
> And just to verify my configuration got setup properly. Can you please verify these settings are OK? Just want to make sure it's setup correctly.
> 
> ~# nextdns config
> control /var/run/nextdns.sock
> config ******
> max-ttl 5s
> cache-size 10MB
> setup-router true
> listen localhost:53
> log-queries false
> detect-captive-portals false
> hardened-privacy false
> bogus-priv true
> timeout 5s
> auto-activate true
> cache-max-age 0s
> report-client-info true
> discovery-dns
> use-hosts true

It’s nothing to be scared of. If you type in the following:

netstat -tulpn | grep :53

you will see nextdns is already binding do to that port but it’s not working. 

![Uploading 5BCB738D-3239-461A-98A9-33B80D01150F.png…]()

**Comment by jbehrmusic** [comment]
> I’ve tried killing the service, restart service, reboot device, uninstall/reinstall. None of them are resolving DNS.

This may be a stupid question, but how can you tell if they are resolving DNS? Because when I go to my my.nextdns.io configuration page, it says "**All good! This device is using NextDNS with this configuration.**", with a green dot next to it. I assume this means my Synology is properly resolving my DNS?

I tried blocking facebook in the 'parental control' tab, and it blocks it on my browser and my phone. So I think my router is picking up the DNS resolve from Synology, correct?

**Comment by netfxtech** [comment]
> > I’ve tried killing the service, restart service, reboot device, uninstall/reinstall. None of them are resolving DNS.
> 
> This may be a stupid question, but how can you tell if they are resolving DNS? Because when I go to my my.nextdns.io configuration page, it says "**All good! This device is using NextDNS with this configuration.**", with a green dot next to it. I assume this means my Synology is properly resolving my DNS?
> 
> I tried blocking facebook in the 'parental control' tab, and it blocks it on my browser and my phone. So I think my router is picking up the DNS resolve from Synology, correct?

Are you using your Synology to resolve DNS or do you have NextDNS installed on your computer?

**Comment by jbehrmusic** [comment]
> Are you using your Synology to resolve DNS or do you have NextDNS installed on your computer?

I have nothing installed on my computer, and I made sure there were no DNS entries in my ethernet adapter. I even ran the ipconfig /flushdns command, then restarted my PC. I also made sure there were no DNS configuration lines in my Firefox settings. I assume this means I am using my Synology NAS to resolve DNS, correct?

I basically did these steps.
1. Uninstalled nextdns on my NAS
2. Restarted my NAS
3. Installed using (https://github.com/nextdns/nextdns/wiki/Synology-DSM)
4. Went into my router (192.168.1.1) and set the Primary DNS to the IP of my Synology NAS (192.1.168.177)
5. SSH into my NAS, ran both these commands: `nextdns starts` & `nextdns activate`

It seems to be working. Although like I said, I still have that weird line after I run `nextdns log` . It shows this: 

2021-02-12T18:39:52-08:00 nextdns[28367]: Startup failed: proxy: tcp: listen tcp :53: bind: address already in use

**Comment by netfxtech** [comment]
Yeah yours is working right now. I have a handful of networks with dns issues because of this right now. If your network quits resolving DNS then you’ll be in good company :)

**Comment by netfxtech** [comment]
It actually appears there are other issues opened for this same problem (referring to the UDM-pro).  Same error.

**Comment by jbehrmusic** [comment]
> It actually appears there are other issues opened for this same problem (referring to the UDM-pro). Same error.

Yeah I saw this issue (bug) opened. 

https://github.com/nextdns/nextdns/issues/413

Not sure if this affects the functionality of resolving the DNS through synology or not. Or how often you have to restart. Maybe the owner/developer can comment on this.

**Comment by adhodgson1** [comment]
I authored issue #413 and in my case Nextdns is running fine on the DSM but the Nextdns status or log is not giving the right output. I did try and troubleshoot this but not sure what is going on. If I need to update Nextdns I have to kill it manually then use the installer to update it, that will get it running. I can also reboot it to get it running as well but any attempt I make to start it via the init script or Nextdns start fails. I am using it with both IPV4 and IPV6 and both clients are working.

**Comment by nicksoda** [comment]
Experienced this as well on a Synology NAS after changing the config for conditional forwarding on a previosuly working installation per the guide(by the config set below).  

Note, I did not use run (used restart as indicated).  The log showed the same bind error (Startup failed: proxy: tcp: listen tcp :53: bind: permission denied), DNS requests were being served, but with the original configuration only.  NextDNS status failed exit code 1.

Uninstall/reinstall nor restart solved the problem. Verified there was nothing binding to port 53 after the uninstall before the reinstall.  I resolved by uninstall, deletion of all the configs from the file, restart the NAS, reinstall, manually edit config file directly for my desired conditional forwarding, NextDNS restart.  I then had a single running instance with proper config and NextDNS properly reporting logs/status. 


Originally changed config to trigger the issue via:

sudo nextdns config set \
    -config 10.0.4.0/24=12345 \
    -config 00:1c:42:2e:60:4a=67890 \
    -config abcdef
sudo nextdns restart
